horovod/spark/common/util.py

```diff
@@ -0,0 +1,518 @@
+        metadata = piece.get_metadata()
+        train_rows += metadata.num_rows
+
+        pfile = piece.open()
+        file_metadata = pfile.metadata
+        for row_group_index in range(file_metadata.num_row_groups):
+            row_group = file_metadata.row_group(row_group_index)
+            total_byte_size += row_group.total_byte_size
+
+    if train_rows == 0:
+        raise ValueError('No rows found in training dataset: {}'.format(train_data_path))
+
+    if total_byte_size == 0:
+        raise ValueError('No data found in training dataset: {}'.format(train_data_path))
+
+    if train_rows > total_byte_size:
+        raise ValueError('Found {} bytes in {} rows.  Training dataset may be corrupted.'
+                         .format(total_byte_size, train_rows))
+
+    val_rows = 0
+    if store.exists(validation_data_path):
+        val_data = pq.ParquetDataset(validation_data_path, filesystem=store.get_filesystem())
+        for piece in val_data.pieces:
+            metadata = piece.get_metadata()
+            val_rows += metadata.num_rows
+
+    schema_cols = feature_columns + label_columns
+    if sample_weight_col:
+        schema_cols.append(sample_weight_col)
+
+    metadata = {}
+    for col in schema_cols:
+        col_schema = schema.field_by_name(col)
+        col_info = {
+            'spark_data_type': pyarrow_to_spark_data_type(col_schema.type),
+            'is_sparse_vector_only': False,
+            'shape': None,  # Only used by SparseVector columns
+            'intermediate_format': constants.NOCHANGE,
+            'max_size': None  # Only used by SparseVector columns
+        }
+        metadata[col] = col_info
+
+    avg_row_size = total_byte_size / train_rows
+    return train_rows, val_rows, metadata, avg_row_size
+
+
+def prepare_data(num_processes, store, df, label_columns, feature_columns,
+                 validation_col=None, validation_split=0.0, sample_weight_col=None,
+                 partitions_per_process=10, verbose=False):
+    if validation_split and validation_col:
+        raise ValueError("can only specify one of validation_col and validation_split")
+
+    num_partitions = num_processes * partitions_per_process
+    if verbose:
+        print('num_partitions={}'.format(num_partitions))
+
+    train_data_path = store.get_train_data_path()
+    val_data_path = store.get_val_data_path()
+    if verbose:
+        print('train_data_path={}'.format(train_data_path))
+        print('val_data_path={}'.format(val_data_path))
+
+    for col in label_columns:
+        if col not in df.columns:
+            raise ValueError('Label column {} does not exist in this DataFrame'.format(col))
+
+    if feature_columns is None:
+        feature_columns = [col for col in df.columns if col not in set(label_columns)]
+
+    dataframe_hash = df.__hash__()
+
+    global _training_cache
+    key = (dataframe_hash, validation_split, validation_col, train_data_path, val_data_path)
+    with _training_cache.lock:
```

Comment:
This looks like a very long lock cycle when cache is cold, this will synchronize multiple threads. Can you please explain why your need to lock the cache anyway? What are the users of the cache / this method?

Reply:
I guess what you are trying to do is to avoid the same traning df is being written to parquet when it is accessed again while it is being created. Hence the per-key lock. This looks like a lazy creation of the parquet files.
Can you initially enumerate all training dataframes that will be used exactly once multi threaded, then you do not need any lock at all.

Reply:
This lock is in place in the event that the user wants to train multiple estimators in parallel without corrupting the cache.  Currently, this would be difficult to do, but we've been working on a system that allows you to do this with a thin API.